File: source/kCura.Relativity.Export.ConvertedToC#/PropertyExtractor/EML.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Data;
using System.Linq;
namespace PropertyExtractor
{
	public class EML
	{

		public System.Collections.Hashtable Extract(string fileName)
		{

			System.IO.StreamReader sr = new System.IO.StreamReader(fileName);
			string lineBuffer = null;
			string header = null;
			string headerValue = null;

			System.Text.StringBuilder body = new System.Text.StringBuilder();
			bool inBody = false;

			System.Collections.Hashtable hashTable = new System.Collections.Hashtable();
			hashTable.Add("File Type".ToLower(), "Email Message");

			lineBuffer = sr.ReadLine();


			while (!(lineBuffer == null)) {

				if (!inBody) {
					if (!string.IsNullOrEmpty(lineBuffer)) {
						if (!string.IsNullOrEmpty(this.ExtractSMTPHeader(lineBuffer))) {
							if (this.IsValidSMTPHeader(this.ExtractSMTPHeader(lineBuffer))) {
								header = this.ExtractSMTPHeader(lineBuffer);
								if (header.ToLower() == "date") {
									headerValue = lineBuffer.Substring(lineBuffer.IndexOf(":") + 1, lineBuffer.IndexOf("-") - lineBuffer.IndexOf(":") - 2);
									try {
										System.DateTime dt = System.DateTime.Parse(headerValue);
									} catch (System.Exception ex) {
										// if the field cannot be parsed then set it as empty
										headerValue = string.Empty;
									}
								} else {
									headerValue = lineBuffer.Substring(lineBuffer.IndexOf(":") + 1);
								}

							} else {
								headerValue = headerValue + lineBuffer;
							}
						} else {
							headerValue = headerValue + lineBuffer;
						}

						if (hashTable.Contains(header.ToLower())) {
							hashTable[header.ToLower()] = headerValue;
						} else {
							hashTable.Add(header.ToLower(), headerValue);
						}

					} else {
						inBody = true;
					}
				} else {
					body.Append(lineBuffer);
					body.Append(Constants.vbCrLf);
				}

				lineBuffer = sr.ReadLine();
			}
			sr.Close();

			hashTable.Add("body", body.ToString());

			return hashTable;

		}

		private string ExtractSMTPHeader(string lineBuffer)
		{
			if (lineBuffer.IndexOf(":") > -1) {
				return lineBuffer.Substring(0, lineBuffer.IndexOf(":"));
			}
			return null;
		}

		private bool IsValidSMTPHeader(string header)
		{
			switch (header.ToLower()) {
				case "to":
				case "bcc":
				case "cc":
				case "message-id":
				case "x-filename":
				case "x-origin":
				case "x-bcc":
				case "from":
				case "x-to":
				case "x-from":
				case "content-type":
				case "content-transfer-Encoding":
				case "x-cc":
				case "x-folder":
				case "mime-version":
				case "subject":
				case "date":
					return true;
				default:
					return false;
			}
		}

	}
}

File: source/kCura.Relativity.Export/PropertyExtractor/EML.vb
Namespace PropertyExtractor
	Public Class EML

		Public Function Extract(ByVal fileName As String) As System.Collections.Hashtable

			Dim sr As New System.IO.StreamReader(fileName)
			Dim lineBuffer As String
			Dim header As String = Nothing
			Dim headerValue As String = Nothing

			Dim body As New System.Text.StringBuilder
			Dim inBody As Boolean

			Dim hashTable As New System.Collections.Hashtable
			hashTable.Add("File Type".ToLower, "Email Message")

			lineBuffer = sr.ReadLine

			Do Until lineBuffer Is Nothing

				If Not inBody Then

					If lineBuffer <> "" Then
						If Me.ExtractSMTPHeader(lineBuffer) <> "" Then
							If Me.IsValidSMTPHeader(Me.ExtractSMTPHeader(lineBuffer)) Then
								header = Me.ExtractSMTPHeader(lineBuffer)
								If header.ToLower = "date" Then
									headerValue = lineBuffer.Substring(lineBuffer.IndexOf(":") + 1, lineBuffer.IndexOf("-") - lineBuffer.IndexOf(":") - 2)
									Try
										Dim dt As Date = Date.Parse(headerValue)
									Catch ex As System.Exception
										' if the field cannot be parsed then set it as empty
										headerValue = String.Empty
									End Try
								Else
									headerValue = lineBuffer.Substring(lineBuffer.IndexOf(":") + 1)
								End If

							Else
								headerValue = headerValue + lineBuffer
							End If
						Else
							headerValue = headerValue + lineBuffer
						End If

						If hashTable.Contains(header.ToLower) Then
							hashTable.Item(header.ToLower) = headerValue
						Else
							hashTable.Add(header.ToLower, headerValue)
						End If

					Else
						inBody = True
					End If
				Else
					body.Append(lineBuffer)
					body.Append(vbCrLf)
				End If

				lineBuffer = sr.ReadLine
			Loop
			sr.Close()

			hashTable.Add("body", body.ToString)

			Return hashTable

		End Function

		Private Function ExtractSMTPHeader(ByVal lineBuffer As String) As String
			If lineBuffer.IndexOf(":") > -1 Then
				Return lineBuffer.Substring(0, lineBuffer.IndexOf(":"))
			End If
			Return Nothing
		End Function

		Private Function IsValidSMTPHeader(ByVal header As String) As Boolean
			Select Case header.ToLower
				Case "to", "bcc", "cc", "message-id", "x-filename", "x-origin", "x-bcc", "from", "x-to", "x-from", "content-type", "content-transfer-Encoding", "x-cc", "x-folder", "mime-version", "subject", "date"
					Return True
				Case Else
					Return False
			End Select
		End Function

	End Class
End Namespace